File: controllers/ImportController.pas
unit ImportController;

interface

procedure Registry;

implementation

uses
  Horse,
  System.JSON,
  System.SysUtils,
  XMLFacturaService,
  ProveedorRepository,
  ProductoRepository,
  ErrorResponseUtils;

procedure PostFacturaXML(Req: THorseRequest; Res: THorseResponse; Next: TProc);
var
  XMLContent: string;
  Factura: TFacturaXML;
  ResponseJSON, ProveedorJSON, ProductoJSON: TJSONObject;
  ProductosArray: TJSONArray;
  I: Integer;
  Proveedor: TProveedorInfo;
  ExisteProductoVar: Boolean;
begin
  try
    XMLContent := Req.Body;
    if XMLContent = '' then
    begin
      SendErrorResponse(Res, 400, 'Cuerpo XML vacío');
      Exit;
    end;

    Factura := TXMLFacturaService.Parsear(XMLContent);

    ResponseJSON := TJSONObject.Create;

    // Validate Provider
    Proveedor := ObtenerProveedorPorNit(Factura.NitProveedor, Factura.Anio);
    ProveedorJSON := TJSONObject.Create;
    ProveedorJSON.AddPair('nit', Factura.NitProveedor);
    ProveedorJSON.AddPair('existe', TJSONBool.Create(Proveedor.Existe));
    if Proveedor.Existe then
      ProveedorJSON.AddPair('codigo', Proveedor.Codigo);
    ResponseJSON.AddPair('proveedor', ProveedorJSON);

    // Validate Products
    ProductosArray := TJSONArray.Create;
    for I := 0 to Length(Factura.Productos) - 1 do
    begin
      ExisteProductoVar := ExisteProducto(Factura.Productos[I].Referencia, Factura.Productos[I].Descripcion, Factura.Anio);

      ProductoJSON := TJSONObject.Create;
      ProductoJSON.AddPair('referencia', Factura.Productos[I].Referencia);
      ProductoJSON.AddPair('descripcion', Factura.Productos[I].Descripcion);
      ProductoJSON.AddPair('existe', TJSONBool.Create(ExisteProductoVar));
      ProductosArray.AddElement(ProductoJSON);
    end;
    ResponseJSON.AddPair('productos', ProductosArray);

    Res.Send(ResponseJSON);
  except
    on E: Exception do
    begin
      LogError(E.Message);
      SendErrorResponse(Res, 500, 'Error interno del servidor', E.Message);
    end;
  end;
end;

procedure Registry;
begin
  THorse.Post('/api/factura/xml', PostFacturaXML);
end;

end.

File: node-backend/src/controllers/importController.js
const { parseInvoice } = require('../utils/xmlParser');
const { obtenerProveedorPorNit } = require('../repositories/proveedorRepository');
const { existeProducto } = require('../repositories/productoRepository');

const postFacturaXml = async (req, res) => {
  const xmlContent = req.body;
  if (!xmlContent) return res.status(400).send('Cuerpo XML vacío');

  const parsed = parseInvoice(xmlContent);
  const anio = String(new Date(parsed.fechaEmision).getFullYear());

  const proveedor = await obtenerProveedorPorNit(parsed.provider.nit, anio);
  const productos = await Promise.all(
    parsed.products.map(async (p) => ({ referencia: p.referencia, descripcion: p.descripcion, existe: await existeProducto(p.referencia, p.descripcion, anio) })),
  );

  return res.json({ proveedor: { nit: parsed.provider.nit, existe: proveedor.existe, ...(proveedor.existe ? { codigo: proveedor.codigo } : {}) }, productos });
};

module.exports = { postFacturaXml };
